File: packages/server/src/repowise/server/schemas/doc_drift.py
"""Documentation drift response models.

The wire shape, and the only place it is declared. Each model is built from a
dict the engine produced rather than from an ORM row, so the route serves what
:mod:`repowise.core.analysis.doc_drift.serialize` emits without a second
hand-maintained mapping in between.

``unavailable`` is a first-class field on both responses and not an HTTP error.
"The drift pass has not run here" is an answer about the index, not a failure
of the request, and a client that received a 4xx would have to guess which of
the three causes it was. An empty list with no ``unavailable`` is the strong
claim that the tree is clean, and only a populated store can support it.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

from repowise.core.analysis.doc_drift.constants import (
    UNAVAILABLE_NO_TABLE,
    UNAVAILABLE_NOT_COMPUTED,
    UNAVAILABLE_READ_FAILED,
)
from repowise.core.analysis.doc_drift.serialize import derive_doc_drift_id

#: Why a drift answer could not be given, from the engine's own vocabulary so
#: the literal cannot be renamed on one side alone.
DocDriftUnavailable = Literal[
    UNAVAILABLE_NOT_COMPUTED,  # type: ignore[valid-type]
    UNAVAILABLE_NO_TABLE,  # type: ignore[valid-type]
    UNAVAILABLE_READ_FAILED,  # type: ignore[valid-type]
]


class DocDriftFindingResponse(BaseModel):
    """One assertion a document makes that the repository no longer satisfies.

    ``file_path`` is the *document* to edit, never the target it names. A
    reader who reads it as the broken file has been told the opposite of the
    truth, which is why the field keeps the name every other drift surface
    gives it.
    """

    #: Stable across stores and rebuilds; see :func:`derive_doc_drift_id`.
    id: str
    file_path: str
    line_number: int
    kind: str
    target: str
    confidence: float
    origin: str
    reason: str
    raw: str
    context: str
    evidence: list[str]

    @classmethod
    def from_dict(cls, data: dict) -> DocDriftFindingResponse:
        """One serialized finding, given the id the store cannot hold."""
        return cls(
            id=derive_doc_drift_id(
                data["file_path"],
                data["kind"],
                data["line_number"],
                data["target"],
            ),
            file_path=data["file_path"],
            line_number=data["line_number"],
            kind=data["kind"],
            target=data["target"],
            confidence=data["confidence"],
            origin=data["origin"],
            reason=data["reason"],
            raw=data["raw"],
            context=data["context"],
            evidence=data.get("evidence", []),
        )


class DocDriftSummaryResponse(BaseModel):
    """The rollup above the list, computed server-side over the same rows.

    ``findings_total`` counts every finding matching the query, including any
    the display cap left out, so a client cannot recompute a page as if it were
    the repository.
    """

    findings_total: int
    documents: int
    confidence: dict[str, int]
    by_kind: dict[str, int]
    #: What the counts do and do not cover. No surface may show them without it.
    findings_basis: str


class DocDriftResponse(BaseModel):
    """Findings for a repository, with the rollup that makes them legible."""

    findings: list[DocDriftFindingResponse]
    #: Rows served after ``limit``; below ``summary.findings_total`` when capped.
    findings_emitted: int
    summary: DocDriftSummaryResponse | None
    unavailable: DocDriftUnavailable | None = None


class DocDriftReferenceResponse(BaseModel):
    """A document that names a file the repository still has.

    Deliberately weaker than a finding: it does not claim the document
    describes the file, or that its prose is current.
    """

    document: str
    line: int
    kind: str
    section: str = ""


class DocDriftDocumentDriftResponse(BaseModel):
    """A naming document that carries drift somewhere in it.

    ``findings`` counts the whole document, not assertions about the file that
    was asked about. A drifted reference resolves to nothing, so no row here
    can say a document's description of that file is wrong.
    """

    document: str
    findings: int


class DocDriftReferencesResponse(BaseModel):
    """Which documents name one file: the reverse view, over HTTP."""

    target_path: str
    references: list[DocDriftReferenceResponse]
    #: Rows served after the cap; below ``references_total`` when capped.
    references_emitted: int
    #: Every reference to the target, so a capped list cannot read as complete.
    references_total: int
    #: Distinct documents naming the target, before any display cap.
    documents: int
    documents_with_drift: list[DocDriftDocumentDriftResponse]
    #: What a reference row claims, and what its absence does not prove.
    references_basis: str
    unavailable: DocDriftUnavailable | None = None

File: packages/types/src/doc-drift.ts
/**
 * Canonical documentation-drift types.
 *
 * Canonical source: the engine's `DocDriftResponse` / `DocDriftReferencesResponse`
 * (`packages/server/src/repowise/server/schemas/doc_drift.py`), emitted verbatim
 * by the route. Nothing here is a normalisation of the wire: a client that
 * reshapes this contract is a client that recomputes the summary, and the
 * summary is served precisely so it cannot be recomputed from a page.
 *
 * The vocabulary carries a claim, and the claims are not the same size:
 * a *finding* says a document asserts something the repository no longer
 * satisfies, and is filed against the document. A *reference* says only that a
 * document names a file the repository still has. It never says the document
 * describes that file.
 */

/**
 * Why a drift answer could not be given.
 *
 * Not an error state. `not_computed` means the drift pass has never run on
 * this index, which must not render as a repository with clean documentation;
 * `index_predates_doc_drift` means the index is older than the store;
 * `drift_read_failed` is anything else, and is separate because "reindex" is
 * the wrong advice for a locked database.
 *
 * Mirrors `UNAVAILABLE_NOT_COMPUTED` / `UNAVAILABLE_NO_TABLE` /
 * `UNAVAILABLE_READ_FAILED` in `core/analysis/doc_drift/constants.py`, pinned
 * by `tests/unit/doc_drift/test_ts_contract_parity.py`.
 */
export type DocDriftUnavailable =
  | "not_computed"
  | "index_predates_doc_drift"
  | "drift_read_failed";

/**
 * The reference classes the detector ships.
 *
 * `symbol` is deliberately absent: it was measured at a 55-69% flag rate and
 * killed, because backticks in technical prose mean "this is a literal token",
 * not "this is a code symbol".
 */
export type DocDriftKind = "path" | "link" | "anchor" | "command";

/**
 * The one set of confidence boundaries, mirroring the engine.
 *
 * `HIGH` is `HIGH_CONFIDENCE_THRESHOLD` and `MEDIUM` is
 * `REVIEW_CONFIDENCE_THRESHOLD` in `core/analysis/doc_drift/constants.py`,
 * which are themselves held equal to dead code's boundaries so that "high"
 * means one thing to a reader comparing the two reports.
 *
 * Duplicated across languages rather than derived, and pinned by a parity
 * test, because three surfaces disagreeing about where "high" starts is a
 * failure this repository has already had once
 * (`tests/unit/dead_code/test_confidence_parity.py`).
 */
export const DOC_DRIFT_CONFIDENCE = {
  /** At or above this, a finding is near-certain. */
  HIGH: 0.7,
  /** The default floor; below it findings are not fetched. */
  MEDIUM: 0.4,
} as const;

/** Which tier a confidence falls in, using the boundaries above. */
export function docDriftConfidenceTier(confidence: number): "high" | "medium" | "low" {
  if (confidence >= DOC_DRIFT_CONFIDENCE.HIGH) return "high";
  if (confidence >= DOC_DRIFT_CONFIDENCE.MEDIUM) return "medium";
  return "low";
}

/**
 * What each reference class is, in a reader's words.
 *
 * The slugs are an API vocabulary, not English: a table column headed `anchor`
 * tells a reader nothing about a heading link whose heading was renamed.
 * Keyed on every value `DocDriftKind` can take, pinned by the same parity test.
 */
export const DOC_DRIFT_KIND_LABELS: Record<DocDriftKind, string> = {
  path: "File path",
  link: "Link",
  anchor: "Heading link",
  command: "Command",
};

/** Label for one reference class, falling back to the raw slug. */
export function docDriftKindLabel(kind: string): string {
  return DOC_DRIFT_KIND_LABELS[kind as DocDriftKind] ?? kind;
}

export interface DocDriftFinding {
  /**
   * Stable across stores and rebuilds, derived from
   * `(file_path, kind, line_number, target)`. The row's own primary key is an
   * autoincrement integer and is re-minted whenever a store is rebuilt, so it
   * is this id that a client keys triage on.
   */
  id: string;
  /**
   * The *document* that is wrong, and the file a reader has to edit. Never the
   * target it names. Rendering this as the broken code file tells the reader
   * the opposite of the truth.
   */
  file_path: string;
  line_number: number;
  kind: string;
  /** What the document claims exists. */
  target: string;
  confidence: number;
  /** Which named resolution strategy produced this. */
  origin: string;
  /** One human sentence. */
  reason: string;
  /** The reference exactly as written. */
  raw: string;
  /** The source line, trimmed, so a finding can be judged without opening the file. */
  context: string;
  /** Lines a reader can check for themselves. */
  evidence: string[];
}

/**
 * The rollup above a findings list, computed server-side over the same rows.
 *
 * `findings_total` counts every finding matching the query, including any the
 * display cap left out, so a client must not recompute these numbers from the
 * page it holds.
 */
export interface DocDriftSummary {
  findings_total: number;
  /** Distinct documents carrying at least one finding. */
  documents: number;
  confidence: Record<string, number>;
  by_kind: Record<string, number>;
  /**
   * What the counts do and do not cover. Most references in a real tree are
   * uncheckable by design, so no surface may show the counts without it.
   */
  findings_basis: string;
}

export interface DocDriftResponse {
  findings: DocDriftFinding[];
  /** Rows served after the cap; below `summary.findings_total` when capped. */
  findings_emitted: number;
  /** Null exactly when `unavailable` is set. */
  summary: DocDriftSummary | null;
  unavailable: DocDriftUnavailable | null;
}

/** A document that names a file the repository still has. */
export interface DocDriftReference {
  document: string;
  line: number;
  kind: string;
  /** The enclosing heading trail, so a reader can find the passage. */
  section?: string;
}

/**
 * A naming document that carries drift somewhere in it.
 *
 * `findings` counts the whole document. A drifted reference resolves to
 * nothing, so this can never mean "this document's description of your file is
 * wrong", and a surface that words it that way is lying.
 */
export interface DocDriftDocumentDrift {
  document: string;
  findings: number;
}

export interface DocDriftReferencesResponse {
  target_path: string;
  references: DocDriftReference[];
  /** Rows served after the cap; below `references_total` when capped. */
  references_emitted: number;
  /** Every reference to the target, so a capped list cannot read as complete. */
  references_total: number;
  /** Distinct documents naming the target, before any display cap. */
  documents: number;
  documents_with_drift: DocDriftDocumentDrift[];
  /** What a reference claims, and what its absence does not prove. */
  references_basis: string;
  unavailable: DocDriftUnavailable | null;
}
